File: app/catalog.py
"""Camada de produtos com duas fontes possíveis por agente:

  * product_mode = "internal": tabela `products` (via ProductRepository).
  * product_mode = "external": API REST do próprio cliente. O backend faz
    GET {product_api_url} (header Authorization: Bearer {product_api_key},
    se configurada) e espera uma lista JSON de produtos
    ([{...}] ou {"products": [...]}). Reserva de estoque não é suportada
    externamente — o agente faz handoff.

Internamente trabalha com ProductRow (tipado); as funções públicas devolvem
dicts prontos para JSON (tools do LLM + respostas da API). Nenhum SQL aqui —
o acesso a dados fica no ProductRepository.
"""

import logging

import requests

from app.domain import AgentConfig, ProductRow
from app.messages import RESERVE_EXTERNAL_UNAVAILABLE
from app.repositories import ProductRepository
from app.textutil import word_set

log = logging.getLogger("blip-agent.catalog")

_EXTERNAL_TIMEOUT = 10  # segundos
_STOPWORDS = {"de", "do", "da", "para", "com", "os", "as", "um", "uma"}

_repo = ProductRepository()


# --------------------------------------------------------------------------- #
# API pública (usada pelas tools e pelos endpoints) — retorna dicts JSON-ready
# --------------------------------------------------------------------------- #

def list_products(agent: AgentConfig) -> list[dict]:
    return [p.to_dict() for p in _fetch_all(agent)]


def search_products(agent: AgentConfig, query: str) -> list[dict]:
    """Busca por substring no nome/descrição; cai para sobreposição de palavras."""
    q = query.lower()
    products = _fetch_all(agent)
    hits = [p for p in products if q in p.name.lower() or q in p.description.lower()]
    if not hits:
        hits = _word_overlap_matches(products, query)
    return [p.to_dict() for p in hits]


def check_stock(agent: AgentConfig, product_name: str, quantity: int) -> dict:
    product = _find_product(agent, product_name)
    if product is None:
        return {"product": product_name, "found": False,
                "error": "Produto não encontrado no catálogo."}
    return {
        "product": product.name,
        "found": True,
        "requested": quantity,
        "available": product.stock,
        "can_fulfill": 0 < quantity <= product.stock,
        "price_unit": round(product.price, 2),
        "price_total": round(product.price * quantity, 2),
        "unit": product.unit,
    }


def reserve_stock(agent: AgentConfig, product_name: str, quantity: int) -> dict:
    """Decrementa estoque (somente modo internal). Externo orienta handoff."""
    if agent.product_mode != "internal":
        return {"success": False, "product": product_name, "error": RESERVE_EXTERNAL_UNAVAILABLE}

    product = _find_product(agent, product_name)
    if product is None:
        return {"success": False, "product": product_name, "error": "Produto não encontrado."}
    if quantity <= 0:
        return {"success": False, "product": product.name, "error": "Quantidade inválida."}

    try:
        updated = _repo.decrement_stock(agent.id, product.id, quantity)
    except ValueError as e:
        return {"success": False, "product": product.name, "error": str(e)}

    log.info("Reservado %d x %r (agent=%s) — novo estoque %d",
             quantity, updated.name, agent.id, updated.stock)
    return {"success": True, "product": updated.name, "new_stock": updated.stock,
            "total_charged": round(product.price * quantity, 2)}


# --------------------------------------------------------------------------- #
# CRUD interno (endpoints de produtos) — delega ao repositório
# --------------------------------------------------------------------------- #

def create_product(agent_id: str, data: dict) -> dict:
    return _repo.create(agent_id, data).to_dict()


def update_product(agent_id: str, product_id: int, changes: dict) -> dict | None:
    product = _repo.update(agent_id, product_id, changes)
    return product.to_dict() if product else None


def delete_product(agent_id: str, product_id: int) -> bool:
    return _repo.delete(agent_id, product_id)


# --------------------------------------------------------------------------- #
# Internos — trabalham com ProductRow
# --------------------------------------------------------------------------- #

def _fetch_all(agent: AgentConfig) -> list[ProductRow]:
    if agent.product_mode == "external":
        return _fetch_external(agent)
    if agent.product_mode == "internal":
        return _repo.list_for_agent(agent.id)
    return []  # product_mode == "none"


def _fetch_external(agent: AgentConfig) -> list[ProductRow]:
    if not agent.product_api_url:
        return []
    headers = {}
    if agent.product_api_key:
        headers["Authorization"] = f"Bearer {agent.product_api_key}"
    try:
        resp = requests.get(agent.product_api_url, headers=headers, timeout=_EXTERNAL_TIMEOUT)
        resp.raise_for_status()
        payload = resp.json()
    except (requests.RequestException, ValueError) as e:
        log.warning("API externa de produtos falhou (agent=%s): %s", agent.id, e)
        return []

    items = payload.get("products", payload) if isinstance(payload, dict) else payload
    if not isinstance(items, list):
        return []
    return [
        ProductRow.from_external(p, fallback_id=i)
        for i, p in enumerate(items)
        if isinstance(p, dict) and p.get("name")
    ]


def _find_product(agent: AgentConfig, product_name: str) -> ProductRow | None:
    """Resolve um produto pelo nome: exato -> substring -> sobreposição de palavras
    (cobre "fones de ouvido" -> "Fone Bluetooth", plurais, ordem trocada)."""
    products = _fetch_all(agent)
    target = product_name.lower()

    for p in products:
        if p.name.lower() == target:
            return p
    substr = [p for p in products if target in p.name.lower()]
    if substr:
        return min(substr, key=lambda p: len(p.name))
    matches = _word_overlap_matches(products, product_name)
    return matches[0] if matches else None


def _word_overlap_matches(products: list[ProductRow], query: str) -> list[ProductRow]:
    """Produtos ordenados por sobreposição de palavras com a query (prefixo conta,
    então "fones" casa com "fone"). Score zero fica de fora."""
    query_words = [w for w in word_set(query) if len(w) >= 3 and w not in _STOPWORDS]
    if not query_words:
        return []
    scored = []
    for p in products:
        hay_words = [w for w in word_set(f"{p.name} {p.description}")
                     if len(w) >= 3 and w not in _STOPWORDS]
        score = sum(1 for qw in query_words
                    if any(hw.startswith(qw) or qw.startswith(hw) for hw in hay_words))
        if score > 0:
            scored.append((score, p))
    scored.sort(key=lambda t: -t[0])
    return [p for _, p in scored]

File: src/server/catalog.ts
/**
 * Catálogo de produtos (porta `app/catalog.py`) com duas fontes por agente:
 *   - product_mode "internal": tabela products (ProductRepository).
 *   - product_mode "external": API REST do cliente (GET product_api_url, com
 *     Authorization: Bearer opcional). Reserva não é suportada externamente.
 *
 * As funções públicas devolvem dados JSON-ready (tools do LLM + endpoints). O
 * modelo NUNCA inventa preço/estoque: tudo vem daqui.
 */

import type { AgentConfig, ProductRow } from "./domain";
import { productRowFromExternal } from "./domain";
import { getLogger } from "./logging";
import { RESERVE_EXTERNAL_UNAVAILABLE } from "./messages";
import { ProductRepository, StockError } from "./repositories/products";
import type { ProductCreateInput } from "./schemas";
import { wordSet } from "./textutil";

const log = getLogger("blip-agent.catalog");

const EXTERNAL_TIMEOUT_MS = 10_000;
const STOPWORDS = new Set(["de", "do", "da", "para", "com", "os", "as", "um", "uma"]);

const repo = new ProductRepository();

// --- API pública (tools + endpoints) — dados JSON-ready ---------------------- //

export async function listProducts(agent: AgentConfig): Promise<ProductRow[]> {
  return fetchAll(agent);
}

export async function searchProducts(agent: AgentConfig, query: string): Promise<ProductRow[]> {
  const q = query.toLowerCase();
  const products = await fetchAll(agent);
  let hits = products.filter((p) => p.name.toLowerCase().includes(q) || p.description.toLowerCase().includes(q));
  if (hits.length === 0) hits = wordOverlapMatches(products, query);
  return hits;
}

export async function checkStock(
  agent: AgentConfig,
  productName: string,
  quantity: number,
): Promise<Record<string, unknown>> {
  const product = await findProduct(agent, productName);
  if (product === null) {
    return { product: productName, found: false, error: "Produto não encontrado no catálogo." };
  }
  return {
    product: product.name,
    found: true,
    requested: quantity,
    available: product.stock,
    can_fulfill: quantity > 0 && quantity <= product.stock,
    price_unit: round2(product.price),
    price_total: round2(product.price * quantity),
    unit: product.unit,
  };
}

export async function catalogHealth(agent: AgentConfig): Promise<Record<string, unknown>> {
  const mode = agent.productMode;
  if (mode === "none") {
    return { mode: "none", configured: false, message: "Catálogo não configurado para este agente." };
  }
  if (mode === "internal") {
    const items = await repo.listForAgent(agent.id);
    return { mode: "internal", configured: true, reachable: true, product_count: items.length };
  }
  // external
  if (!agent.externalProducts) {
    return {
      mode: "external",
      configured: Boolean(agent.productApiUrl),
      enabled: false,
      reachable: false,
      message: "Catálogo externo desligado por feature flag.",
    };
  }
  if (!agent.productApiUrl) {
    return {
      mode: "external",
      configured: false,
      enabled: true,
      reachable: false,
      message: "URL do catálogo externo não configurada.",
    };
  }
  const [reachable, count, detail] = await probeExternal(agent);
  return { mode: "external", configured: true, enabled: true, reachable, product_count: count, detail };
}

export async function reserveStock(
  agent: AgentConfig,
  productName: string,
  quantity: number,
): Promise<Record<string, unknown>> {
  if (agent.productMode !== "internal") {
    return { success: false, product: productName, error: RESERVE_EXTERNAL_UNAVAILABLE };
  }
  const product = await findProduct(agent, productName);
  if (product === null) {
    return { success: false, product: productName, error: "Produto não encontrado." };
  }
  if (quantity <= 0) {
    return { success: false, product: product.name, error: "Quantidade inválida." };
  }
  let updated: ProductRow;
  try {
    updated = await repo.decrementStock(agent.id, product.id, quantity);
  } catch (e) {
    if (e instanceof StockError) return { success: false, product: product.name, error: e.message };
    throw e;
  }
  log.info(`Reservado ${quantity} x ${updated.name} (agent=${agent.id}) — novo estoque ${updated.stock}`);
  return {
    success: true,
    product: updated.name,
    new_stock: updated.stock,
    total_charged: round2(product.price * quantity),
  };
}

// --- CRUD interno (endpoints de produtos) ----------------------------------- //

export async function createProduct(agentId: string, data: ProductCreateInput): Promise<ProductRow> {
  return repo.create(agentId, data);
}

export async function updateProduct(
  agentId: string,
  productId: number,
  changes: Partial<ProductCreateInput>,
): Promise<ProductRow | null> {
  return repo.update(agentId, productId, changes);
}

export async function deleteProduct(agentId: string, productId: number): Promise<boolean> {
  return repo.delete(agentId, productId);
}

// --- Internos --------------------------------------------------------------- //

async function fetchAll(agent: AgentConfig): Promise<ProductRow[]> {
  if (agent.productMode === "external") {
    if (!agent.externalProducts) {
      log.warn(`Catálogo externo desligado por flag (agent=${agent.id}).`);
      return [];
    }
    return fetchExternal(agent);
  }
  if (agent.productMode === "internal") return repo.listForAgent(agent.id);
  return [];
}

function externalHeaders(agent: AgentConfig): HeadersInit {
  return agent.productApiKey ? { Authorization: `Bearer ${agent.productApiKey}` } : {};
}

async function externalGet(agent: AgentConfig): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), EXTERNAL_TIMEOUT_MS);
  try {
    const resp = await fetch(agent.productApiUrl, { headers: externalHeaders(agent), signal: controller.signal });
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return await resp.json();
  } finally {
    clearTimeout(timer);
  }
}

function extractItems(payload: unknown): unknown[] {
  if (Array.isArray(payload)) return payload;
  if (payload && typeof payload === "object" && Array.isArray((payload as Record<string, unknown>).products)) {
    return (payload as Record<string, unknown>).products as unknown[];
  }
  return [];
}

async function fetchExternal(agent: AgentConfig): Promise<ProductRow[]> {
  if (!agent.productApiUrl) return [];
  let payload: unknown;
  try {
    payload = await externalGet(agent);
  } catch (e) {
    log.warn(`API externa de produtos falhou (agent=${agent.id}):`, String(e));
    return [];
  }
  return extractItems(payload)
    .filter((p): p is Record<string, unknown> => !!p && typeof p === "object" && Boolean((p as Record<string, unknown>).name))
    .map((p, i) => productRowFromExternal(p, i));
}

async function probeExternal(agent: AgentConfig): Promise<[boolean, number, string]> {
  let payload: unknown;
  try {
    payload = await externalGet(agent);
  } catch (e) {
    return [false, 0, `Falha ao acessar o catálogo externo: ${String(e)}`];
  }
  return [true, extractItems(payload).length, "OK"];
}

/** Resolve um produto pelo nome: exato -> substring -> sobreposição de palavras. */
async function findProduct(agent: AgentConfig, productName: string): Promise<ProductRow | null> {
  const products = await fetchAll(agent);
  const target = productName.toLowerCase();

  for (const p of products) {
    if (p.name.toLowerCase() === target) return p;
  }
  const substr = products.filter((p) => p.name.toLowerCase().includes(target));
  if (substr.length > 0) {
    return substr.reduce((best, p) => (p.name.length < best.name.length ? p : best));
  }
  const matches = wordOverlapMatches(products, productName);
  return matches[0] ?? null;
}

/** Produtos ordenados por sobreposição de palavras com a query (prefixo conta). */
function wordOverlapMatches(products: ProductRow[], query: string): ProductRow[] {
  const queryWords = [...wordSet(query)].filter((w) => w.length >= 3 && !STOPWORDS.has(w));
  if (queryWords.length === 0) return [];
  const scored: Array<[number, ProductRow]> = [];
  for (const p of products) {
    const hayWords = [...wordSet(`${p.name} ${p.description}`)].filter((w) => w.length >= 3 && !STOPWORDS.has(w));
    const score = queryWords.reduce(
      (acc, qw) => acc + (hayWords.some((hw) => hw.startsWith(qw) || qw.startsWith(hw)) ? 1 : 0),
      0,
    );
    if (score > 0) scored.push([score, p]);
  }
  scored.sort((a, b) => b[0] - a[0]);
  return scored.map(([, p]) => p);
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}
